Source/DafnyCore/Verifier/ProofObligationDescription.cs

```diff
@@ -1827,9 +1855,106 @@ internal class Utils {
     return new BinaryExpr(lowRange.tok, BinaryExpr.Opcode.Or, lowRange, highRange);
   }
 
+  public static void MakeQuantifierVarsForDims(List<Expression> dims, out List<BoundVar> vars, out List<Expression> varExprs, out Expression range) {
+    var zero = new LiteralExpr(Token.NoToken, 0);
+    vars = dims.Select((_, i) => new BoundVar(Token.NoToken, "i" + i, Type.Int)).ToList();
+
+    // can't assign to out-param immediately, since it's accessed in the lambda below
+    var tempVarExprs = vars.Select(var => new IdentifierExpr(Token.NoToken, var) as Expression).ToList();
+    var indexRanges = dims.Select((dim, i) => new ChainingExpression(
+      Token.NoToken,
+      new() { zero, tempVarExprs[i], dim },
+      new() { BinaryExpr.Opcode.Le, BinaryExpr.Opcode.Lt },
+      new() { Token.NoToken, Token.NoToken },
+      new() { null, null }
+    ) as Expression).ToList();
+    varExprs = tempVarExprs;
+
+    range = dims.Count == 1 ? indexRanges[0] : new ChainingExpression(
+      Token.NoToken,
+      indexRanges,
+      Enumerable.Repeat(BinaryExpr.Opcode.And, dims.Count - 1).ToList(),
+      Enumerable.Repeat(Token.NoToken as IToken, dims.Count - 1).ToList(),
+      Enumerable.Repeat(null as Expression, dims.Count - 1).ToList()
+    );
+  }
+
   internal static Expression MakeIsOneCtorAssertion(Expression root, List<DatatypeCtor> ctors) {
     return ctors
       .Select(ctor => new ExprDotName(Token.NoToken, root, ctor.Name + "?", null) as Expression)
       .Aggregate((e0, e1) => new BinaryExpr(Token.NoToken, BinaryExpr.Opcode.Or, e0, e1));
   }
+
+  /// <summary>
+  /// Builds an expression that represents whether (the relevant subset of) the given `supersetFrames`
+  /// permit read/modification access to all objects, arrays, and/or sets of objects/arrays in the `subsetFrames`.
+  /// </summary>
+  public static Expression MakeDafnyMultiFrameCheck(List<FrameExpression> supersetFrames, List<FrameExpression> subsetFrames) {
```

Comment:
The detail required in methods like these make me very much want to share this code with the Boogie translation eventually (i.e., have the Boogie generator work by constructing this expression and then translating it). But that's a task for another day.